setup.sh

```diff
@@ -1,126 +1,223 @@
+# supported version of docker.
+function addDockerYumRepo() {
+   tee /etc/yum.repos.d/docker.repo <<-'EOF'
 [dockerrepo]
 name=Docker Repository
 baseurl=https://yum.dockerproject.org/repo/main/centos/7/
 enabled=1
 gpgcheck=1
 gpgkey=https://yum.dockerproject.org/gpg
 EOF
+}
+
+# Set up yum and install the supported version of docker
+function installDocker() {
+   addDockerYumRepo
 
-yum -y install docker-engine-1.10.3
+   yum -y install docker-engine-1.10.3
+}
 
-# Set docker daemon comand line options. Keep in mind that at this point this
+# Set docker daemon comand line options. We modify systemd configuration
+# for docker to start with our desired options.
+# Keep in mind that at this point this
 # overrides any existing options supplied by the RPM. This is overridden to
 # make sure docker is listening on all network interfaces.
-echo "" > /etc/sysconfig/docker
-mkdir /etc/systemd/system/docker.service.d
-tee /etc/systemd/system/docker.service.d/docker.conf <<-'EOF'
+function setDockerDaemonOptions() {
+   echo "" > /etc/sysconfig/docker
+   mkdir -p /etc/systemd/system/docker.service.d
+   tee /etc/systemd/system/docker.service.d/docker.conf <<-'EOF'
 [Service]
 ExecStart=
 ExecStart=/usr/bin/docker daemon --selinux-enabled -H unix:///var/run/docker.sock -H tcp://0.0.0.0:2375
 EOF
-# Start docker.
-systemctl start docker
-# Start docker on startup.
-systemctl enable docker
-echo "Complete."
-
-
-GOVERSION=1.6.2
-GOBINARY=go${GOVERSION}.linux-amd64.tar.gz
-echo "Installing go ${GOVERSION}..."
-wget -q https://storage.googleapis.com/golang/$GOBINARY
-tar -C /usr/local/ -xzf $GOBINARY
-ln -sf /usr/local/go/bin/* /usr/bin/
-rm -f $GOBINARY
-echo "Complete."
-
-
-#ETCDVERSION=v2.2.2
-#echo "Installing etcd ${ETCDVERSION}..."
-#ETCDNAME=etcd-${ETCDVERSION}-linux-amd64
-#ETCDBINARY=${ETCDNAME}.tar.gz
-#wget -q https://github.com/coreos/etcd/releases/download/${ETCDVERSION}/${ETCDBINARY}
-#tar -C /usr/local/ -xzf ${ETCDBINARY}
-#mv /usr/local/${ETCDNAME} /usr/local/etcd
-#ln -s /usr/local/etcd/etcd /usr/bin/etcd
-#ln -s /usr/local/etcd/etcd-migrate /usr/bin/etcd-migrate
-#ln -s /usr/local/etcd/etcdctl /usr/bin/etcdctl
-#rm $ETCDBINARY
-#echo "Complete."
-
-
-echo "Creating a GOPATH in /home/vagrant/gopath local to the VM..."
-# Create a gopath and symlink in the src directory. (Since we don't want to
-# share bin/ and pkg/ since they are platform dependent.)
-mkdir -p /home/vagrant/gopath/bin /home/vagrant/gopath/pkg
-ln -s $GOPATH/src /home/vagrant/gopath/src
-echo "Complete."
-
-
+}
+
+# startDocker starts the docker service using systemctl
+function startDocker() {
+
+   setDockerDaemonOptions
+
+   systemctl daemon-reload
+   systemctl start docker
+   systemctl enable docker
+   echo "Docker daemon started."
+}
+
+# Downloads the file with wget if it does not exist in the current directory.
+# The user passes the wget argument path to this function as the first parameter
+function maybeDownloadFile() {
```

Comment:
maybe_download_file to me says that it's almost random whether it will or not. ensure_file_is_downloaded works for me.